docs/public/specifications/tests/DNSSEC-TP/dnssec03.md

```diff
@@ -1,73 +1,316 @@
+and interpreted ([RFC 5155][RFC 5155#section-3], section3):
+
+* Hash algorithm
+* Flags
+* Iterations
+* Salt
+
+**Hash algorithm:** The only legal value of the hash algorithm is value 1
+(SHA-1). See ([RFC 5155][RFC 5155#section-11], section 11 and
+[IANA NSEC3 Parameters registry]).
+
+**Flags:** The only defined flags in the flag field is bit 7, "opt-out". It may
+only be set in the NSEC record, not in the NSEC3PARAM record
+([RFC 5155][RFC 5155#section-11], section 11 and
+[IANA NSEC3 Parameters registry]). "For small zones, the use of opt-out-based
+NSEC3 records is NOT RECOMMENDED. For very large and sparsely signed zones, where
+the majority of the records are insecure delegations, opt-out MAY be used"
+([RFC 9276][RFC 9276#section-3.1], section 3.1). This means that unless the zone
+is a TLD or a TLD like domain found in the [Public Suffix List] it should
+not have the opt-out bit set.
+
+**Iterations:** For a name server an increased number of NSEC3 iterations have a
+negative impact on performance. The recommendation is to have 0 iterations. "If
+NSEC3 must be used, then an iterations count of 0 MUST be used to alleviate
+computational burdens" ([RFC 9276][RFC 9276#section-3.1], section 3.1).
+
+**Salt:** The salt parameter has been seen as a security feature but
+[RFC 9276][RFC 9276#section-3.1], section 3.1, states that zones "SHOULD NOT use
+a salt by indicating a zero-length salt value instead". The justification for
+the recommendation is found in [RFC 9276][RFC 9276#section-2.4], section 2.4.
+
+
+## Scope
+
+This test case will just ignore non-responsive name servers or name servers not
+giving a correct DNS response for an authoritative name server (covered by
+[Connectivity01]).
+
+This test case is only relevant if the zone has been DNSSEC signed.
+
+
+## Inputs
+
+* "Child Zone" - The domain name to be tested.
+* "Public Suffix List Data" - The list or a copy of the list found at
+  [Public Suffix List data].
+
+
+## Summary
+
+* If no DNSKEY records are found, no further investigation will be done.
 
-### Objective
+Message Tag outputted              | Level   | Arguments  | Message ID for message tag
+:----------------------------------|:--------|:-----------|:--------------------------------------------
+DS03_ERR_MULT_NSEC3                | ERROR   | ns_list | Multiple NSEC3 records when one is expected. Fetched from name servers "{ns_list}".
+DS03_ILLEGAL_HASH_ALGO             | ERRRO   | ns_list, algo_num | The following servers respond with an illegal hash algorithm for NSEC3 ({algo_num}). Fetched from name servers "{ns_list}".
+DS03_ILLEGAL_ITERATION_VALUE       | ERROR   | ns_list, int | The following servers respond with the NSEC3 iteration value {int} (not recommended). Fetched from name servers "{ns_list}".
+DS03_ILLEGAL_SALT_LENGTH           | WARNING | ns_list, int | The following servers respond with a non-empty salt in NSEC3 ({int} octets). Fetched from name servers "{ns_list}".
```

Comment:
The current recommendation could maybe be made clearer in those messages. For example:
```suggestion
DS03_ILLEGAL_ITERATION_VALUE       | ERROR   | ns_list, int | The following servers respond with the NSEC3 iteration value {int}. The recommended practice is to set this value to 0. Fetched from name servers "{ns_list}".
DS03_ILLEGAL_SALT_LENGTH           | WARNING | ns_list, int | The following servers respond with a non-empty salt in NSEC3 ({int} octets). The recommended practice is to use an empty salt. Fetched from name servers "{ns_list}".
```